### PR TITLE
feat: filesystem-event watcher for external file changes

### LINKED ISSUE #558: External file changes are only detected via the periodic git-pull reindex
## Summary

The server detects external file changes only via the periodic git-pull loop's synchronous reindex. Operators without git configured, or with `git_pull_interval_s=0`, get no automatic detection of files changed on disk outside the server; B-tool answers can be silently stale for an unbounded window.

## Observed [verified: read src/markdown_vault_mcp/_server_deps.py, collection.py, _server_tools.py, config.py at HEAD e076017]

External file changes — file modifications, additions, or deletions made on disk by anything other than the MCP server's `write`/`edit`/`delete`/`rename` tools — are detected only by the periodic git-pull loop. When a pull occurs (per `git_pull_interval_s`, default 600s), the loop's `on_pull=reindex` callback runs `Collection.reindex()`, which calls `ChangeTracker.detect_changes` to SHA-scan the working tree.

When the operator has not configured git (`git_pull_interval_s=0`, or no `git_strategy` configured), this loop does not start. There is no other periodic or background detection. The only triggers for `reindex()` are:

- Lifespan-start (only on cold start; warm restart short-circuits via the completeness sentinel without scanning disk).
- The git-pull loop (only if configured).
- Operator-triggered `reindex` MCP tool or CLI `index` command.

Between these triggers, B-tool answers reflect the FTS state at the last detection, regardless of what's currently on disk.

## Context

- Vault directories touched by an external tool (text editor, file sync daemon, OS file restore, `cp -r`, etc.) without going through the MCP server or git pull.
- Operators running without git integration entirely (private vaults stored as plain directories).
- The current default `git_pull_interval_s=600` means even with git polling enabled, the steady-state detection window is 10 minutes.

## Scope boundary

- Not about #534 (drift handling on B3 read paths once drift is known). #534 takes the server-known drift state as input; this issue is about populating that state when changes happen outside the server's control.
- Not about reacting to detected changes (block-catch-up vs warn-and-stale); that contract is #534's once drift is known.

## Open questions

- [unverified] Whether the project's deployment model (Docker, headless servers) reliably supports OS-level file-watcher APIs (inotify on Linux, FSEvents on macOS, ReadDirectoryChangesW on Windows) inside typical container contexts.
- [unverified] Whether operator workflows that bypass both MCP-write-tools and git pulls are common enough to warrant default-on detection vs. opt-in.

## Relationship

Sibling of #534, #535, #536. May inform or be informed by #534's drift-aware reader design.

### PR DESCRIPTION
**Pull request by pvliesdonk**
## Summary

- Adds `VaultFileWatcher` backed by `watchdog` (inotify on Linux, FSEvents on macOS) that triggers `reindex()` after a debounce window when files are modified by external tools (text editors, sync daemons, `cp -r`, etc.)
- Fixes the gap identified in #558: non-git vaults and git vaults with `GIT_PULL_INTERVAL_S=0` had no automatic detection of external file changes
- Mutually exclusive with the git pull loop and GitHub webhook: both already trigger reindex on their own cadence; mixing them would cause mid-checkout partial scans

## Design

- **`_file_watcher.py`** — `VaultFileWatcher` with debounced `on_change` callback; `_VaultEventHandler` filters hidden-directory events (`.git/`, `.markdown_vault_mcp/`); graceful no-op when watchdog is not installed
- **Hidden-dir filtering** — `PurePosixPath('.').parts` returns `()` (empty), not `('.',)`, so `DirModifiedEvent` on source_dir itself needs an explicit empty-parts guard (debugged via live watchdog event trace)
- **Auto-disable rule** — lifespan checks `git_pull_interval_s > 0 or github_webhook_secret`; logs INFO when disabled
- **Graceful degradation** — `_WATCHDOG_AVAILABLE` flag; `start()` logs WARNING and returns cleanly if watchdog is not installed
- **Optional dep** — `markdown-vault-mcp[file-watcher]` = `watchdog>=4.0`; added to dev deps so CI always has it

## New config

| Variable | Default | Description |
|---|---|---|
| `MARKDOWN_VAULT_MCP_FILE_WATCHER` | `true` | Enable watcher; auto-disabled with git pull/webhook |
| `MARKDOWN_VAULT_MCP_FILE_WATCHER_DEBOUNCE_S` | `2.0` | Quiet window before reindex fires |

## Test plan

- [x] `test_file_change_triggers_on_change` — real inotify event → on_change after debounce
- [x] `test_rapid_changes_trigger_single_on_change` — 10 writes in 100ms → 1 call
- [x] `test_stop_before_debounce_cancels_callback` — stop() within window → no call
- [x] `test_stop_is_idempotent` — double stop() no exception
- [x] `test_hidden_dir_changes_are_ignored` — `.git/COMMIT_EDITMSG` → no call
- [x] `test_nested_hidden_dir_changes_are_ignored` — `.markdown_vault_mcp/state/index.db` → no call
- [x] `test_visible_file_after_hidden_change_triggers_callback` — visible write after hidden → call
- [x] `test_start_logs_warning_when_watchdog_unavailable` — mock unavailable → no raise
- [x] 4 auto-disable logic tests covering no-git-active, git-pull-active, webhook-active, explicitly-disabled
- [x] 1810 total tests passing

Closes #558

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/markdown_vault_mcp/_file_watcher.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+      on source_dir fires for every file addition/deletion inside it, hidden or
+      not.  Filtering it out is safe because the concrete file events always
+      accompany it and carry the actual path information.
+    """
+    from pathlib import PurePosixPath
+
+    parts = PurePosixPath(path).parts
+    if not parts:
+        return True
+    return any(part.startswith(".") for part in parts)
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/markdown_vault_mcp/_file_watcher.py]: Fixed in f017e10. `_has_hidden_component` now takes `tuple[str, ...]` and the call site passes `rel.parts` directly — the platform-native `Path.relative_to()` already splits on the correct separator, so no `PurePosixPath` round-trip needed. Verified the fix handles the `visible/subdir/.git/file` case that the original string-based approach missed on Windows.

File: src/markdown_vault_mcp/_file_watcher.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            self._source_dir = source_dir
+
+        def on_any_event(self, event: FileSystemEvent) -> None:
+            path = getattr(event, "src_path", "") or ""
+            try:
+                from pathlib import Path as _Path
+
+                rel = _Path(path).relative_to(self._source_dir)
+                if _has_hidden_component(str(rel)):
+                    return
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/markdown_vault_mcp/_file_watcher.py]: Fixed in f017e10 — `_has_hidden_component(rel.parts)` at the call site.

File: src/markdown_vault_mcp/_file_watcher.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        source_dir: Path,
+        on_change: Callable[[], None],
+        debounce_s: float = 2.0,
+    ) -> None:
+        self._source_dir = source_dir
+        self._on_change = on_change
+        self._debounce_s = debounce_s
+        self._observer: Any = None
+        self._timer: threading.Timer | None = None
+        self._lock = threading.Lock()
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/markdown_vault_mcp/_file_watcher.py:116]: Fixed in f017e10. `_stopped = False` added to `__init__`.

File: src/markdown_vault_mcp/_file_watcher.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+    def _schedule(self) -> None:
+        """Reset the debounce timer."""
+        with self._lock:
+            if self._timer is not None:
+                self._timer.cancel()
+            self._timer = threading.Timer(self._debounce_s, self._fire)
+            self._timer.daemon = True
+            self._timer.start()
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/markdown_vault_mcp/_file_watcher.py:128]: Fixed in f017e10. `_schedule()` now returns early inside the lock if `_stopped` is True.

File: src/markdown_vault_mcp/_file_watcher.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+    def _fire(self) -> None:
+        with self._lock:
+            self._timer = None
+        try:
+            self._on_change()
+        except Exception:
+            logger.error("file_watcher: on_change callback raised", exc_info=True)
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/markdown_vault_mcp/_file_watcher.py:138]: Fixed in f017e10. `_fire()` checks `_stopped` inside the lock before clearing `self._timer` and before invoking `on_change`.

File: src/markdown_vault_mcp/_file_watcher.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    def start(self) -> None:
+        """Start watching *source_dir*.  No-op when watchdog is not installed."""
+        if not _WATCHDOG_AVAILABLE:
+            logger.warning(
+                "file_watcher: watchdog not installed; external file changes "
+                "will not trigger automatic reindex. "
+                "Install watchdog: pip install 'markdown-vault-mcp[file-watcher]'"
+            )
+            return
+
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/markdown_vault_mcp/_file_watcher.py:154]: Fixed in f017e10. `start()` resets `_stopped = False` inside the lock before creating the observer, allowing safe restarts.

File: src/markdown_vault_mcp/_file_watcher.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+                self._timer = None
+
+        if self._observer is not None:
+            try:
+                self._observer.stop()
+                self._observer.join(timeout=5.0)
+            except Exception:
+                logger.warning("file_watcher: error stopping observer", exc_info=True)
+            finally:
+                self._observer = None
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/markdown_vault_mcp/_file_watcher.py]: Fixed in f017e10. `stop()` now sets `_stopped = True` and atomically extracts the observer reference in a single lock acquisition: `observer, self._observer = self._observer, None`. The observer is then stopped and joined outside the lock. Also added the double-call guard in `start()` (`if self._observer is not None: return`).

**Comment by pvliesdonk** [comment]
Addressed the second-round review findings across `ca7b644` and `6365c21`:

**1. Theoretical `start()` race** — `self._observer` is now assigned (under lock) *before* `observer.start()`, with an exception guard that clears it if `start()` throws. A racing `stop()` can now always see and stop the observer.

**2. Redundant `import time as _time`** — removed; uses the module-level `time`.

**3. `test_start_logs_warning_when_watchdog_unavailable` didn't assert the warning** — now uses `caplog` and asserts `"watchdog not installed"` is emitted.

**4. `_server_deps.py` wiring untested** — added three integration tests that drive `make_collection_lifespan` directly via `async with lifespan_fn(None)` and assert `VaultFileWatcher.start`/`stop` are (not) called for the no-git, git-pull-active, and webhook-active configs.

**That fourth one earned its keep — it caught a real bug.** `should_start_file_watcher` was reading `config.git_pull_interval_s`, which defaults to **600 even on non-git vaults**. So on every plain non-git vault the watcher concluded "git pull is active" and never started — the default-on file watcher would in fact never run in production. The isolated logic tests passed because they hardcoded the interval; only driving the real lifespan exposed it.

Fix: `should_start_file_watcher` now takes `git_pull_active: bool` (so the ambiguous interval→active inference can't be made by a caller again), and the lifespan computes it from the *resolved* `kwargs["git_pull_interval_s"]` — the same value `to_collection_kwargs()` hands the Collection, which is 0 unless a git strategy is configured.

1827 tests pass; diff-cover 85%.
